src/Hazelcast.Net.Caching/HazelcastCache.cs

```diff
@@ -0,0 +1,168 @@
+﻿// Copyright (c) 2008-2022, Hazelcast, Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+// http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Hazelcast.Configuration;
+using Hazelcast.Core;
+using Hazelcast.DistributedObjects;
+using Microsoft.Extensions.Caching.Distributed;
+
+namespace Hazelcast.Caching;
+
+/// <summary>
+/// Provides a <see cref="IDistributedCache"/> implementation based on Hazelcast.
+/// </summary>
+/// <remarks>
+/// <para>The <see cref="HazelcastCache"/> relies on a <see cref="IHazelcastClient"/> which
+/// needs to be properly configured, and in particular MUST be configured for reconnecting
+/// the cluster (see <see cref="Networking.NetworkingOptions.ReconnectMode"/>). Otherwise,
+/// the cache may end up in a disconnected state, and never be able to recover.</para>
+/// </remarks>
+public class HazelcastCache : IDistributedCache, IAsyncDisposable
```

Comment:
WDYT about sealed to devirtualize?

Reply:
That class is inherited by `ProvidedHazelcastClass` and `ProvidedHazelcastCacheWithFailover` so we cannot seal it, OTOH we can seal those two classes.